pkg/virt-launcher/virtwrap/api/schema.go

```diff
@@ -661,29 +662,50 @@ type InterfaceTarget struct {
 }
 
 type Alias struct {
-	Name string `xml:"name,attr"`
+	name        string
+	userDefined bool
 }
 
-type UserAlias Alias
+// Package private, responsible to interact with xml.Encoder and xml.Decoder
+type userAliasXML struct {
+	Name string `xml:"name,attr"`
+}
 
 type Rom struct {
 	Enabled string `xml:"enabled,attr"`
 }
 
+func NewAlias(aliasName string) *Alias {
+	return &Alias{name: aliasName}
+}
+
+func NewUserDefinedAlias(aliasName string) *Alias {
+	return &Alias{name: aliasName, userDefined: true}
+}
+
+func (alias Alias) GetName() string {
+	return alias.name
+}
+
 func (alias Alias) MarshalXML(e *xml.Encoder, start xml.StartElement) error {
-	userAlias := UserAlias(alias)
-	userAlias.Name = UserAliasPrefix + userAlias.Name
-	return e.EncodeElement(userAlias, start)
+	userAliasXml := userAliasXML{Name: alias.name}
+	if alias.userDefined {
+		userAliasXml.Name = UserAliasPrefix + userAliasXml.Name
+	}
+	return e.EncodeElement(userAliasXml, start)
```

Comment:
I am not 100% sure that the `MarshalXML` and `UnmarshaXML` bits work properly. Maybe spend a little bit more time on the unit tests for it. They may be responsible for the clear e2e test error.

Reply:
Will go through the e2e failures. As quick note:
- `e2e-k8s-1.17/18/19` : 1 failure in `watch_test`
- `e2e-k8s-1.17-root-ceph/cnao-1.19` : 17 failures
- `windows1026` : 1 failure in `windows_test`

Reply:
It really looks like the mapping in virt-handler does not work anymore. Very likely due to broken serialization and deserialization, so that it can't find the disks and interfaces anymore. Ensure first that the xml parts work as expected. There is a high chance that this then resolves all the errors.

Reply:
Will put more test to it. The updated version looks in sync with the previous version. Specifically:

- Marshal : 
  - make-a-copy-of-the-alias, 
  - (alter-the-copy-if-needed), 
  - encode-the-copy

- Unmarshal: 
  - decode-into-a-var, 
  - make-a-copy-of-the-var, 
  - (alter-the-copy-if-needed), 
  - have the alias point to the copy.

Reply:
As of now, I could not find the leakage. With trials, I found followings:

1. If I replace the `name` in Alias type with any variable name starting with a capital letter (i.e. exported/public/visible-outside-of-api), function tests all pass.
2. If I replace the `name` in Alias type with any other package-private/internal  name, the `watch_test` for 'kubectl get vmi` would fail for not showing the IP.

So, I'd like to propose to take this fix for now. I will in the meanwhile work to find out the leakage.

Reply:
As stated in other responses. The issue was because the JSON marshaling/unmarshaling for the communication between the `virt-handler` and `virt-launcher` for the package-private members of `Alias` objects would also require special treatment. This was addressed by adding two functions, `MarshalJSON` and `UnmarshalJSON`, for `Alias` objects.